{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#1. Define The Problem\n",
    "\n",
    "When we do PCR to amplify our libraries, we will end up with many copies of reads. Using certain signatures found in a SAM file we can identify these PCR duplicates and eliminate them. If we do not do this, certain analyses, such as gene expression, can be altered by the presence of the same read over and over again as it will just be interpretted as a very abundant transcript, thus messing up the analysis.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#2. Write examples\n",
    "\n",
    "Example Input:\n",
    "\n",
    "1:497:R:-272+13M17D24M:ATCGTTG\t113\t1\t497\t37\t37M\t15\t100338662\t0\tCGGGTCTGACCTGAGGAGAACTGTGCTCCGCCTTCAG\t0;==-==9;>>>>>=>>>>>>>>>>>=>>>>>>>>>>\tXT:A:U\tNM:i:0\tSM:i:37\tAM:i:0\tX0:i:1\tX1:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tMD:Z:37\n",
    "\n",
    "\n",
    "19:20389:F:275+18M2D19M:ATCGTTG\t99\t1\t497\t0\t37M\t=\t17919\t314\tCGAGTCTGACCTGAGGAGAAGTGTGCTCCGCATTCAG\t>>>>>>>>>>>>>>>>>>>><<>>><<>>4::>>:<9\tRG:Z:UM0098:1\tXT:A:R\tNM:i:0\tSM:i:0\tAM:i:0\tX0:i:4\tX1:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tMD:Z:37\n",
    "\n",
    "\n",
    "19:20389:F:275+18M2D19M:ATCGGTG\t147\t1\t499\t0\t2S35M\t=\t17644\t-314\tTTGGTCTGACCTGAGGAGAACTGTGCTCCGCCTTCAG\t;44999;499<8<8<<<8<<><<<<><7<;<<<>><<\tXT:A:R\tNM:i:2\tSM:i:0\tAM:i:0\tX0:i:4\tX1:i:0\tXM:i:0\tXO:i:1\tXG:i:2\tMD:Z:18^CA19\n",
    "\n",
    "\n",
    "9:21597+10M2I25M:R:-209:TTGAACT\t83\t1\t21678\t0\t8M2I27M\t=\t21469\t-244\tCACCACATCACATATACCAAGCCTGGCTGTGTCTTCT\t<;9<<5><<<<><<<>><<><>><9>><>>>9>>><>\tXT:A:R\tNM:i:2\tSM:i:0\tAM:i:0\tX0:i:5\tX1:i:0\tXM:i:0\tXO:i:1\tXG:i:2\tMD:Z:35\n",
    "\n",
    "Example Output:\n",
    "\n",
    "1:497:R:-272+13M17D24M\t113\t1\t497\t37\t37M\t15\t100338662\t0\tCGGGTCTGACCTGAGGAGAACTGTGCTCCGCCTTCAG\t0;==-==9;>>>>>=>>>>>>>>>>>=>>>>>>>>>>\tXT:A:U\tNM:i:0\tSM:i:37\tAM:i:0\tX0:i:1\tX1:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tMD:Z:37\n",
    "\n",
    "9:21597+10M2I25M:R:-209\t83\t1\t21678\t0\t8M2I27M\t=\t21469\t-244\tCACCACATCACATATACCAAGCCTGGCTGTGTCTTCT\t<;9<<5><<<<><<<>><<><>><9>><>>>9>>><>\tXT:A:R\tNM:i:2\tSM:i:0\tAM:i:0\tX0:i:5\tX1:i:0\tXM:i:0\tXO:i:1\tXG:i:2\tMD:Z:35\n",
    "\n",
    "\n",
    "We got rid of the second and third read because the second read has almost the same sequence, the same start position, and the same UMI as the first read, thus it is a PCR duplicate, and the third read has the same read after the first two bases and has soft clipping that accounts for those first two being off, as well as having a very similar UMI, so it is also a PCR duplicate. The fourth read is completely different so it is not a PCR duplicate and will remain."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#3. Pseudocode\n",
    "\n",
    "Notes:\n",
    "\n",
    "- Want bitwise flag to determine strandedness\n",
    "- don't need sequnce\n",
    "- need corrected start position, UMI, and strand\n",
    "- If reverse strand want to check end position (technically where that strand started)\n",
    "- For paired end reads, need to a double check that both pairs are not PCR duplicates\n",
    "- Look at sam file to see how much you need to save at a time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Run samtools sort on SAM file and feed ouput to following Python script\n",
    "\n",
    "#Initiate an empty dictionary\n",
    "#Open SAM file\n",
    "    #Use readline to read each line of the file one at a time\n",
    "    #run isolate_UMI to look specifically at the UMI\n",
    "    #If we have a list of known UMIs check if UMI is in list and toss line if it is not\n",
    "    #Check if UMI is in dictionary\n",
    "        #If no, add line from file to dictionary as a value, where the key is the UMI\n",
    "            #write this line to an output file that will contain no PCR duplicates\n",
    "        #If yes, run check_PCR_dup on all lines in the dictionary that are values for that key\n",
    "            #If check_PCR_dup is false then it is not a PCR duplicate\n",
    "                #We then want to add this file line as a value to the UMI key already in the dictionary\n",
    "                #Also write this line to the same output file \n",
    "            #If check_PCR_dup is true then it is a PCR duplicate\n",
    "                #Can write duplicates to a separate output file if so desired\n",
    "                #Now we want to ignore the PCR duplicate so we just move on to the next line of the file\n",
    "    #Now we need to flush the dictionary every so often so it doesn't get too big \n",
    "    #To make sure we get all of the PCR duplicates, we don't want to flush the dictionary too early\n",
    "    #Create a variable that saves the length of the line from the file\n",
    "    #Rewrite this variable if the line you are reading is longer than the previously saved variable\n",
    "    #If the start position of a line is 2 times this saved variable away from the previously saved lines in the dictionary,\n",
    "    #flush the dictionary because since the file is sorted by start position PCR duplicates cannot possibly occur that \n",
    "    #far away from each other.\n",
    "    \n",
    "    #Stop reading through the file once it has been fully read through\n",
    "    \n",
    "    #For paired-end reads, do the same thing except look at 2 lines of file at a time because the paired end file is formatted\n",
    "    #such that there is a read and then its read pair directly after it\n",
    "    #For paired reads to be PCR duplicates, both the read and read pair have to match up with a previously discovered read, \n",
    "    #read-pair combo."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#4. High Level Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def start_pos_check(string1,string2):\n",
    "    '''Checks the start position of two reads seeing if they're matching'''\n",
    "    #Isolate column containing start position\n",
    "    #Check cigar string for soft clipping\n",
    "    #If there is soft clipping, subtract amount of soft clipping from start position \n",
    "    #After adjusting for soft clipping, see if start position matches\n",
    "    #Return True if match, False if no match\n",
    "    #Example input: string 1 start = 100 CIGAR = 100M, string 2 start = 110 CIGAR = 10S90M\n",
    "    #Example output: True\n",
    "    \n",
    "def isolate_UMI(string1, string2):\n",
    "    '''Checks if UMIs are matching'''\n",
    "    #Use regex to isolate umi from qname\n",
    "    #Return UMI\n",
    "    #Example input: 1:497:R:-272+13M17D24M:ATCGTTG\n",
    "    #Example ouput: ATCGTTG\n",
    "    \n",
    "def check_chrom(string1, string2):\n",
    "    '''Checks if read comes from the same chromosome'''\n",
    "    #Use regex to isolate chromosome from rname\n",
    "    #Checks to see if from same chromosome\n",
    "    #Return True if match, False if no match\n",
    "    #Example input: 1:497:R:-272+13M17D24M:ATCGTTG\t113\t1\t497\t37\t37M\t15\t100338662\t0\tCGGGTCTGACCTGAGGAGAACTGTGCTCCGCCTTCAG\t0;==-==9;>>>>>=>>>>>>>>>>>=>>>>>>>>>>\tXT:A:U\tNM:i:0\tSM:i:37\tAM:i:0\tX0:i:1\tX1:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tMD:Z:37\n",
    "    #Example input: 1:497:R:-272+13M17D24M:ATCGTTG\t113\t1\t497\t37\t37M\t15\t100338662\t0\tCGGGTCTGACCTGAGGAGAACTGTGCTCCGCCTTCAG\t0;==-==9;>>>>>=>>>>>>>>>>>=>>>>>>>>>>\tXT:A:U\tNM:i:0\tSM:i:37\tAM:i:0\tX0:i:1\tX1:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tMD:Z:37\n",
    "    #Example output: True (because the RNAME for both is 1)\n",
    "\n",
    "def end_pos_check(string1,string2):\n",
    "    '''Checks end postion of two reads to see if they're matching'''\n",
    "    #Isolate column containing start position\n",
    "    #Check cigar string for soft clipping\n",
    "    #If there is soft clipping, subtract amount of soft clipping from start position \n",
    "    #Now add the number next to the M to the start position to find the end position\n",
    "    #Return True if match, False is no match\n",
    "    #Example input: Line with start position of 100 CIGAR string of 100M\n",
    "    #Example input: Line with start position of 100 CIGAR sting of 10S80M10S\n",
    "    #Example output: False (input 1 has end position 200, input 2 has end position 190)\n",
    "    \n",
    "def strand_check(string):\n",
    "    '''Checks if reads came from same strand of DNA'''\n",
    "    #Isolate column containing bitwise flag\n",
    "    #Decode bitwise flag to look for strandedness flag \n",
    "    #Return \"+\" if forward, \"-\" if reverse.\n",
    "    #Example Input: 1:497:R:-272+13M17D24M:ATCGTTG\t113\t1\t497\t37\t37M\t15\t100338662\t0\tCGGGTCTGACCTGAGGAGAACTGTGCTCCGCCTTCAG\t0;==-==9;>>>>>=>>>>>>>>>>>=>>>>>>>>>>\tXT:A:U\tNM:i:0\tSM:i:37\tAM:i:0\tX0:i:1\tX1:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tMD:Z:37\n",
    "    #Example Input: 1:497:R:-272+13M17D24M:ATCGTTG\t113\t1\t497\t37\t37M\t15\t100338662\t0\tCGGGTCTGACCTGAGGAGAACTGTGCTCCGCCTTCAG\t0;==-==9;>>>>>=>>>>>>>>>>>=>>>>>>>>>>\tXT:A:U\tNM:i:0\tSM:i:37\tAM:i:0\tX0:i:1\tX1:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tMD:Z:37\n",
    "    #Example Output: True (both are +)\n",
    "    \n",
    "def check_PCR_dup(string1,string2):\n",
    "    '''Checks if a PCR duplicate is present. Returns True if they are PCR duplicates, False if they are not'''\n",
    "    if check_chrom(string1,string2) == True:\n",
    "        if strand_check(string1)==strand_check(string2):\n",
    "            if strand_check(string1) == \"+\":\n",
    "                if start_pos_check(string1,string2) == True:\n",
    "                    return True\n",
    "                else:\n",
    "                    return False\n",
    "            if strand_check(string1) == \"-\":\n",
    "                if end_pos_check(string1,string2) == True:\n",
    "                    return True\n",
    "                else:\n",
    "                    return False\n",
    "        else:\n",
    "            return False\n",
    "    else:\n",
    "        return False\n",
    "    \n",
    "    #Example Input: 1:497:R:-272+13M17D24M:ATCGTTG\t113\t1\t497\t37\t37M\t15\t100338662\t0\tCGGGTCTGACCTGAGGAGAACTGTGCTCCGCCTTCAG\t0;==-==9;>>>>>=>>>>>>>>>>>=>>>>>>>>>>\tXT:A:U\tNM:i:0\tSM:i:37\tAM:i:0\tX0:i:1\tX1:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tMD:Z:37\n",
    "    #Example Input: 1:497:R:-272+13M17D24M:ATCGTTG\t113\t1\t497\t37\t37M\t15\t100338662\t0\tCGGGTCTGACCTGAGGAGAACTGTGCTCCGCCTTCAG\t0;==-==9;>>>>>=>>>>>>>>>>>=>>>>>>>>>>\tXT:A:U\tNM:i:0\tSM:i:37\tAM:i:0\tX0:i:1\tX1:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tMD:Z:37\n",
    "    #Example Output: True\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
